File: SFNetworkOscillator/SFNetworkOscillator.cs
using Diploma2.Networks;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;

namespace Diploma2
{
    [Serializable]
    public class SFNetworkOscillatorState
    {
        public double Time { get; set; }
        public double[] Phases { get; set; }
        public SFNetworkOscillatorState(double time, double[] phases)
        {
            Time = time;
            Phases = phases;
        }
        public override string ToString()
        {
            return "[" + Time + ": " + Phases.Length + " phases]";
        }

        public static SFNetworkOscillatorState Debinarize(BinaryReader br)
        {
            double time;
            int phaseCount;
            time = BitConverter.ToDouble(br.ReadBytes(8), 0);
            phaseCount = BitConverter.ToInt32(br.ReadBytes(4), 0);
            List<double> phases = new List<double>();
            for (int i = 0; i < phaseCount; i++)
                phases.Add(BitConverter.ToDouble(br.ReadBytes(8), 0));
            return new SFNetworkOscillatorState(time, phases.ToArray());
        }
    }
    [Serializable]
    public class SFNetworkOscillator : SFNetwork
    {
        private const double pi2 = Math.PI * 2;

        public double Strength { get; private set; }
        public double FrequenciesInitMin { get; private set; }
        public double FrequenciesInitMax { get; private set; }
        public double PhasesInitMin { get; private set; }
        public double PhasesInitMax { get; private set; }
        public double TimeInit { get; private set; }
        public double TimeStep { get; private set; }
        public double SolveStep { get; private set; }

        public double Time { get; private set; }

        ///// Oscillators /////
        public double[] Frequencies { get; private set; }
        public double[] Phases { get; private set; }
        public double[] PhasesInit { get; private set; } // left for compability
        public List<SFNetworkOscillatorState> States { get; private set; }
        ///// Oscillators /////

        private RK4S.RK4SFunc[] funcs = null;

        public SFNetworkOscillator(int node_count, int mlt, double strength, double freqMin, double freqMax, double phaseMin, double phaseMax, double time_init, double time_step, double solve_step) : base(node_count, mlt)
        {
            constructor(strength, time_init, freqMin, freqMax, phaseMin, phaseMax, time_step, solve_step);
        }
        public SFNetworkOscillator(int node_count, int mlt, double strength, double freqMin, double freqMax, double phaseMin, double phaseMax, double time_init, double time_step, double solve_step, int seed) : base(node_count, mlt, seed)
        {
            constructor(strength, time_init, freqMin, freqMax, phaseMin, phaseMax, time_step, solve_step);
        }
        private void constructor(double strength, double time_init, double freqMin, double freqMax, double phaseMin, double phaseMax, double time_step, double solve_step)
        {
            Strength = strength;
            FrequenciesInitMin = freqMin;
            FrequenciesInitMax = freqMax;
            PhasesInitMin = phaseMin;
            PhasesInitMax = phaseMax;
            TimeInit = time_init;
            Time = time_init;
            TimeStep = time_step;
            SolveStep = solve_step;

            Frequencies = new double[Nodes.Count];
            Phases = new double[Nodes.Count];
            for (int i = 0; i < Nodes.Count; i++)
            {
                Frequencies[i] = ilRandom.Next((int)freqMin, (int)freqMax);
                Phases[i] = ilRandom.NextDouble(phaseMin, phaseMax);
            }
            States = new List<SFNetworkOscillatorState> { new SFNetworkOscillatorState(Time, Phases) };
            constructorFuncs();
        }
        private void constructorFuncs()
        {
            funcs = new RK4S.RK4SFunc[Nodes.Count];
            for (int i = 0; i < funcs.Length; i++)
                funcs[i] = (ind, t, x) =>
                {
                    double res = 0;
                    for (int j = 0; j < x.Length; j++) // Foreach phase
                    {
                        //if (j == ind) continue; // Skip if checking itself
                        int w = 0; // w_ij
                        foreach (Edge edg in Edges)
                            if (edg.From == j && edg.To == ind || edg.From == ind && edg.To == j)
                            {
                                w += edg.Weight;
                                break;
                            }

                        // Iterative part of main equation
                        res += w * Math.Sin(x[j] - x[ind]);
                    }
                    // Main equation
                    return Phases[ind] + Strength * res;
                };
        }

        public void SimulateDynamicStep()
        {
            RK4S.RK4SResult result = RK4S.Solve(funcs, Time, Phases, Time + TimeStep, SolveStep);
            Time = result.Time;
            Phases = result.Values;
            phasesNormalize();
            States.Add(new SFNetworkOscillatorState(Time, Phases));
        }
        
        public new int Serialize(string path)
        {
            try
            {
                BinaryFormatter bf = new BinaryFormatter();
                using (FileStream fs = File.OpenWrite(path))
                    bf.Serialize(fs, this);
            }
            catch
            {
                return 1;
            }
            return 0;
        }
        public static new SFNetworkOscillator Deserialize(string path)
        {
            try
            {
                BinaryFormatter bf = new BinaryFormatter();
                using (FileStream fs = File.OpenRead(path))
                    return (SFNetworkOscillator)bf.Deserialize(fs);
            }
            catch { }
            return null;
        }

        private void phasesNormalize()
        {
            for (int i = 0; i < Phases.Length; i++)
                Phases[i] = Phases[i] - (int)(Phases[i] / pi2) * pi2;
        }

        public SFNetworkOscillator(int nodeCount, int mlt, int seed, ulong x, Edge[] edges, double strength, double freqMin, double freqMax, double phaseMin, double phaseMax, double time_init, double time_step, double solve_step, double time, double[] phases, double[] freqs, SFNetworkOscillatorState[] states)
            : base(nodeCount, mlt, edges, new ilRand(seed, x))
        {
            Strength = strength;
            FrequenciesInitMin = freqMin;
            FrequenciesInitMax = freqMax;
            PhasesInitMin = phaseMin;
            PhasesInitMax = phaseMax;
            TimeInit = time_init;
            Time = time;
            TimeStep = time_step;
            SolveStep = solve_step;
            Phases = phases;
            Frequencies = freqs;
            States = new List<SFNetworkOscillatorState>(states);
            constructorFuncs();
        }
        public static SFNetworkOscillator Debinarize(string path)
        {
            try
            {
                int version, nodeCount, mlt, seed, edgeCount, phaseCount, freqCount, stateCount;
                ulong x;
                double strength,
                    freq_init_min,
                    freq_init_max,
                    phase_init_min,
                    phase_init_max,
                    time_init,
                    time_step,
                    solve_step,

                    time;
                using (BinaryReader br = new BinaryReader(File.OpenRead(path)))
                {
                    version = br.ReadBytes(1)[0];
                    nodeCount = BitConverter.ToInt32(br.ReadBytes(4), 0);
                    mlt = BitConverter.ToInt32(br.ReadBytes(4), 0);
                    seed = BitConverter.ToInt32(br.ReadBytes(4), 0);
                    x = (ulong)BitConverter.ToInt64(br.ReadBytes(8), 0);

                    edgeCount = BitConverter.ToInt32(br.ReadBytes(4), 0);
                    List<Edge> edges = new List<Edge>();
                    for (int i = 0; i < edgeCount; i++)
                        edges.Add(Edge.Debinarize(br));

                    strength = BitConverter.ToDouble(br.ReadBytes(8), 0);
                    freq_init_min = BitConverter.ToDouble(br.ReadBytes(8), 0);
                    freq_init_max = BitConverter.ToDouble(br.ReadBytes(8), 0);
                    phase_init_min = BitConverter.ToDouble(br.ReadBytes(8), 0);
                    phase_init_max = BitConverter.ToDouble(br.ReadBytes(8), 0);
                    time_init = BitConverter.ToDouble(br.ReadBytes(8), 0);
                    time_step = BitConverter.ToDouble(br.ReadBytes(8), 0);
                    solve_step = BitConverter.ToDouble(br.ReadBytes(8), 0);
                    time = BitConverter.ToDouble(br.ReadBytes(8), 0);

                    phaseCount = BitConverter.ToInt32(br.ReadBytes(4), 0);
                    List<double> phases = new List<double>();
                    for (int i = 0; i < phaseCount; i++)
                        phases.Add(BitConverter.ToDouble(br.ReadBytes(8), 0));

                    freqCount = BitConverter.ToInt32(br.ReadBytes(4), 0);
                    List<double> freqs = new List<double>();
                    for (int i = 0; i < freqCount; i++)
                        freqs.Add(BitConverter.ToDouble(br.ReadBytes(8), 0));

                    stateCount = BitConverter.ToInt32(br.ReadBytes(4), 0);
                    List<SFNetworkOscillatorState> states = new List<SFNetworkOscillatorState>();
                    for (int i = 0; i < stateCount; i++)
                        states.Add(SFNetworkOscillatorState.Debinarize(br));

                    return new SFNetworkOscillator(nodeCount, mlt, seed, x, edges.ToArray(), strength, freq_init_min, freq_init_max, phase_init_min, phase_init_max, time_init, time_step, solve_step, time, phases.ToArray(), freqs.ToArray(), states.ToArray());
                }
            }
            catch (Exception e) { return null; }
        }
    }
}

File: CPP_SFNetworkOscillator/SFNetworkOscillator.cpp
#include "SFNetwork.h"
#include "ilRand.h"
#include "SFNetworkOscillator.h"

#include <vector>
#include <algorithm>
#include <random>
#include <fstream>


#define PI2				6.283185307179586476925286766559


namespace Diploma2 {


	SFNetworkOscillatorState::SFNetworkOscillatorState() {}
	SFNetworkOscillatorState::SFNetworkOscillatorState(double t, std::vector<double> p) {
		time = t;
		phases = p;
	}
	SFNetworkOscillatorState::~SFNetworkOscillatorState() {};
	void SFNetworkOscillatorState::Binarize(std::ofstream &str) {
		str.write((char*)&time, sizeof(time));

		int phase_count = phases.size();
		str.write((char*)&phase_count, sizeof(phase_count));
		for (double ph : phases)
			str.write((char*)&ph, sizeof(ph));
	}
	SFNetworkOscillatorState SFNetworkOscillatorState::Debinarize(std::ifstream & str)
	{
		double t;
		int p_cnt;
		str.read(reinterpret_cast<char *>(&t), sizeof(t));

		str.read(reinterpret_cast<char *>(&p_cnt), sizeof(p_cnt));
		std::vector<double> ph(p_cnt);
		for (int i = 0; i < p_cnt; i++)
			str.read(reinterpret_cast<char *>(&ph[i]), sizeof(ph[i]));

		return SFNetworkOscillatorState(t, ph);
	}



	void SFNetworkOscillator::constructor_funcs() {
		funcs = std::vector<RK4SFunc>(node_count);
		for (int i = 0; i < funcs.size(); i++) {
			funcs[i] = [this](int ind, double t, const std::vector<double>& x)->double {
				double res = 0;
				for (int j = 0; j < x.size(); j++)
				{
					//if (j == ind) continue;
					int w = 0;
					for (Edge edg : this->edges) {
						if (edg.from == j && edg.to == ind || edg.from == ind && edg.to == j)
						{
							w += edg.weight;
							break;
						}
					}
					res += w * sin(x[j] - x[ind]);
				}
				return phases[ind] + strength * res;
			};
		}
	}
	void SFNetworkOscillator::constructor(double str, double f_min, double f_max, double p_min, double p_max, double t_init, double t_step, double s_step) {
		strength = str;
		freq_init_min = f_min;
		freq_init_max = f_max;
		phase_init_min = p_min;
		phase_init_max = p_max;
		time_init = t_init;
		time = t_init;
		time_step = t_step;
		solve_step = s_step;

		freqs = std::vector<double>(node_count);
		phases = std::vector<double>(node_count);
		for (int i = 0; i < node_count; i++)
		{
			freqs[i] = ilRandom.Next((int)f_min, (int)f_max);
			phases[i] = ilRandom.NextDouble(p_min, p_max);
		}
		states = std::vector<SFNetworkOscillatorState>{ SFNetworkOscillatorState(time, phases) };
		constructor_funcs();
	}

	SFNetworkOscillator::SFNetworkOscillator() : SFNetwork() {}
	SFNetworkOscillator::SFNetworkOscillator(int node_count, int mlt, double strength, double freq_min, double freq_max, double phase_min, double phase_max, double time_init, double time_step, double solve_step
	) {
		initialize(node_count, mlt, strength, freq_min, freq_max, phase_min, phase_max, time_init, time_step, solve_step);
	}
	SFNetworkOscillator::SFNetworkOscillator(int node_count, int mlt, double strength, double freq_min, double freq_max, double phase_min, double phase_max, double time_init, double time_step, double solve_step, int random_seed
	) {
		initialize(node_count, mlt, strength, freq_min, freq_max, phase_min, phase_max, time_init, time_step, solve_step, random_seed);
	}
	SFNetworkOscillator::SFNetworkOscillator(int nodeCount, int mlt, int seed, ulong x, std::vector<Edge> &edges, double str, double fmin, double fmax, double pmin, double pmax, double ti, double ts, double ss, double t, std::vector<double> &ph, std::vector<double> &fr, std::vector<SFNetworkOscillatorState> &st
	) {
		initialize(nodeCount, mlt, seed, x, edges, str, fmin, fmax, pmin, pmax, ti, ts, ss, t, ph, fr, st);
	}
	SFNetworkOscillator::SFNetworkOscillator(const std::string &path) {
		initialize(path);
	}
	SFNetworkOscillator::~SFNetworkOscillator() {};



	void SFNetworkOscillator::initialize(int node_count, int mlt, double strength, double freq_min, double freq_max, double phase_min, double phase_max, double time_init, double time_step, double solve_step) {
		initialize_sfnw(node_count, mlt);
		constructor(
			strength,
			freq_min,
			freq_max,
			phase_min,
			phase_max,
			time_init,
			time_step,
			solve_step
		);
	}
	void SFNetworkOscillator::initialize(int node_count, int mlt, double strength, double freq_min, double freq_max, double phase_min, double phase_max, double time_init, double time_step, double solve_step, int random_seed) {
		initialize_sfnw(node_count, mlt, random_seed);
		constructor(
			strength,
			freq_min,
			freq_max,
			phase_min,
			phase_max,
			time_init,
			time_step,
			solve_step
		);
	}
	void SFNetworkOscillator::initialize(int nodeCount, int mlt, int seed, ulong x, std::vector<Edge> &edges, double str, double fmin, double fmax, double pmin, double pmax, double ti, double ts, double ss, double t, std::vector<double> &ph, std::vector<double> &fr, std::vector<SFNetworkOscillatorState> &st) {
		initialize_sfnw(nodeCount, mlt, edges, ilRand(seed, x));

		strength = str;
		freq_init_min = fmin;
		freq_init_max = fmax;
		phase_init_min = pmin;
		phase_init_max = pmax;
		time_init = ti;
		time_step = ts;
		solve_step = ss;
		time = t;
		phases = ph;
		freqs = fr;
		states = st;

		constructor_funcs();
	}
	void SFNetworkOscillator::initialize(const std::string &path) {
		int n, m, s, e_cnt, p_cnt, f_cnt, s_cnt;
		ulong x;
		double stre,
			fmin,
			fmax,
			pmin,
			pmax,
			ti,
			ts,
			ss,
			t;

		std::ifstream str(path, std::ios::binary);
		char version = 0;
		str.read(reinterpret_cast<char *>(&version), sizeof(version));

		str.read(reinterpret_cast<char *>(&n), sizeof(n));
		str.read(reinterpret_cast<char *>(&m), sizeof(m));
		str.read(reinterpret_cast<char *>(&s), sizeof(s));
		str.read(reinterpret_cast<char *>(&x), sizeof(long long));

		str.read(reinterpret_cast<char *>(&e_cnt), sizeof(e_cnt));
		std::vector<Edge> edgs(e_cnt);
		for (int i = 0; i < e_cnt; i++)
			edgs[i] = Edge::Debinarize(str);

		str.read(reinterpret_cast<char *>(&stre), sizeof(stre));
		str.read(reinterpret_cast<char *>(&fmin), sizeof(fmin));
		str.read(reinterpret_cast<char *>(&fmax), sizeof(fmax));
		str.read(reinterpret_cast<char *>(&pmin), sizeof(pmin));
		str.read(reinterpret_cast<char *>(&pmax), sizeof(pmax));
		str.read(reinterpret_cast<char *>(&ti), sizeof(ti));
		str.read(reinterpret_cast<char *>(&ts), sizeof(ts));
		str.read(reinterpret_cast<char *>(&ss), sizeof(ss));
		str.read(reinterpret_cast<char *>(&t), sizeof(t));

		str.read(reinterpret_cast<char *>(&p_cnt), sizeof(p_cnt));
		std::vector<double> ph(p_cnt);
		for (int i = 0; i < p_cnt; i++)
			str.read(reinterpret_cast<char *>(&ph[i]), sizeof(ph[i]));

		str.read(reinterpret_cast<char *>(&f_cnt), sizeof(f_cnt));
		std::vector<double> fr(f_cnt);
		for (int i = 0; i < f_cnt; i++)
			str.read(reinterpret_cast<char *>(&fr[i]), sizeof(fr[i]));

		str.read(reinterpret_cast<char *>(&s_cnt), sizeof(s_cnt));
		std::vector<SFNetworkOscillatorState> st(s_cnt);
		for (int i = 0; i < s_cnt; i++)
			st[i] = SFNetworkOscillatorState::Debinarize(str);

		initialize(n, m, s, x, edgs, stre, fmin, fmax, pmin, pmax, ti, ts, ss, t, ph, fr, st);
	}


	void SFNetworkOscillator::phasesNormalize() {
		for (int i = 0; i < phases.size(); i++)
		{
			phases[i] = phases[i] - (int)(phases[i] / PI2) * PI2;
			if (phases[i] < 0) phases[i] = PI2 + phases[i];
		}
	}
	void SFNetworkOscillator::SimulateDynamicStep() {
		RK4SResult result = RK4S::Solve(funcs, time, phases, time + time_step, solve_step);
		time = result.time;
		phases = result.values;
		phasesNormalize();
		states.push_back(SFNetworkOscillatorState(time, phases));
	}


	void SFNetworkOscillator::Binarize(const char* path, const char version) {
		switch (version)
		{
		case 1:
			Binarize_v1(path); break;
		default: break;
		}
	}
	void SFNetworkOscillator::Binarize_v1(const char* path) {
		std::ofstream ofile(path, std::ios::binary);
		char version = 1;
		ofile.write(&version, sizeof(version));
		ofile.write((char*)&node_count, sizeof(node_count));
		ofile.write((char*)&multiplier, sizeof(multiplier));
		ofile.write((char*)&seed, sizeof(seed));
		ofile.write((char*)&ilRandom.x, sizeof(ilRandom.x));
		int edge_count = edges.size();
		ofile.write((char*)&edge_count, sizeof(edge_count));
		for (Edge edg : edges)
			edg.Binarize(ofile);

		ofile.write((char*)&strength, sizeof(strength));
		ofile.write((char*)&freq_init_min, sizeof(freq_init_min));
		ofile.write((char*)&freq_init_max, sizeof(freq_init_max));
		ofile.write((char*)&phase_init_min, sizeof(phase_init_min));
		ofile.write((char*)&phase_init_max, sizeof(phase_init_max));
		ofile.write((char*)&time_init, sizeof(time_init));
		ofile.write((char*)&time_step, sizeof(time_step));
		ofile.write((char*)&solve_step, sizeof(solve_step));
		ofile.write((char*)&time, sizeof(time));

		int phase_count = phases.size();
		ofile.write((char*)&phase_count, sizeof(phase_count));
		for (double ph : phases)
			ofile.write((char*)&ph, sizeof(ph));

		int freq_count = freqs.size();
		ofile.write((char*)&freq_count, sizeof(freq_count));
		for (double fr : freqs)
			ofile.write((char*)&fr, sizeof(fr));

		int state_count = states.size();
		ofile.write((char*)&state_count, sizeof(state_count));
		for (SFNetworkOscillatorState st : states)
			st.Binarize(ofile);
	}
}